crypto/rand/drbg_lib.c

```diff
@@ -229,6 +239,103 @@ int RAND_DRBG_reseed(RAND_DRBG *drbg,
 }
 
 /*
+ * Restart |drbg|, using the specified entropy or additional input
+ *
+ * Tries its best to get the drbg instantiated by all means,
+ * regardless of its current state.
+ *
+ * Optionally, a |buffer| of |len| random bytes can be passed,
+ * which is assumed to contain at least |entropy| bits of entropy.
+ *
+ * If |entropy| > 0, the buffer content is used as entropy input.
+ *
+ * If |entropy| == 0, the buffer content is used as additional input
+ *
+ * This function is used internally only.
+ */
+int rand_drbg_restart(RAND_DRBG *drbg,
+                      const unsigned char *buffer, size_t len, size_t entropy)
+{
+    int reseeded = 0;
+
+    const unsigned char *adin = NULL;
+    size_t adinlen = 0;
+
+    if (drbg->pool != NULL) {
+        RANDerr(RAND_F_RAND_DRBG_RESTART, ERR_R_INTERNAL_ERROR);
+        RAND_POOL_free(drbg->pool);
+        drbg->pool = NULL;
+        return 0;
+    }
+
+    if (buffer != NULL) {
+        if (entropy > 0) {
+            if (drbg->max_entropylen < len) {
+                RANDerr(RAND_F_RAND_DRBG_RESTART,
+                    RAND_R_ENTROPY_INPUT_TOO_LONG);
+                return 0;
+            }
+
+            if (entropy > 8 * len) {
+                RANDerr(RAND_F_RAND_DRBG_RESTART, RAND_R_ENTROPY_OUT_OF_RANGE);
+                return 0;
+            }
+
+            /* will be picked up by the rand_drbg_get_entropy() callback */
+            drbg->pool = RAND_POOL_new(entropy, len, len);
+            RAND_POOL_add(drbg->pool, buffer, len, entropy);
+        } else {
+            if (drbg->max_adinlen < len) {
+                RANDerr(RAND_F_RAND_DRBG_RESTART,
+                        RAND_R_ADDITIONAL_INPUT_TOO_LONG);
+                return 0;
+            }
+            adin = buffer;
+            adinlen = len;
+        }
+    }
+
+    /* repair error state */
+    if (drbg->state == DRBG_ERROR)
+        RAND_DRBG_uninstantiate(drbg);
+
+    /* repair uninitialized state */
+    if (drbg->state == DRBG_UNINITIALISED) {
+        drbg_setup(drbg, NULL);
+        /* already reseeded. prevent second reseeding below */
+        reseeded = (drbg->state == DRBG_READY);
+    }
+
+    /* refresh current state if entropy or additional input has been provided */
+    if (drbg->state == DRBG_READY) {
+        if (adin != NULL) {
+            /*
+             * mix in additional input without reseeding
+             *
+             * Similar to RAND_DRBG_reseed(), but the provided additional
+             * data |adin| is mixed into the current state without pulling
+             * entropy from the trusted entropy source using get_entropy().
+             * This is not a reseeding in the strict sense of NIST SP 800-90A.
+             */
+            ctr_reseed(drbg, adin, adinlen, NULL, 0);
+        } else if (reseeded == 0) {
+            /* do a full reseeding if it has not been done yet above */
+            RAND_DRBG_reseed(drbg, adin, adinlen);
```

Comment:
You are certainly giving me a harder time than Columbo could :wink:.

I'm not sure whether I understand what you mean with "obfuscate the spelling of 'NULL, 0' ".  
Maybe it is because you noticed that `adin` and `adinlen` are passed as arguments 2 and 3
```C
     ctr_reseed(drbg, adin, adinlen, NULL, 0);
```
and not arguments 4 and 5, as they should be?
```C
     ctr_reseed(drbg, NULL, 0, adin, adinlen);
```
An obvious reason why I did this is that `ctr_reseed()` fails if `entropy == NULL`,
```C
int ctr_reseed(RAND_DRBG *drbg,
               const unsigned char *entropy, size_t entropylen,
               const unsigned char *adin, size_t adinlen)
{
==> if (entropy == NULL)
==>     return 0;
    ctr_update(drbg, entropy, entropylen, adin, adinlen, NULL, 0);
    return 1;
}
```
nevertheless passing `adin` as `entropy` looks like a dirty hack and you might have wondered why I didn't just call
```C
    ctr_update(drbg, NULL, 0, adin, adinlen, NULL, 0);
```
Initially, this was my intention, but then I hesitated when I noticed that in the `ctr_update()` function the `in1` and `in2` buffers are not treated in exactly the same way: The counter gets updated only if `in1len != 0`:
```C
static void ctr_update(RAND_DRBG *drbg,
                       const unsigned char *in1, size_t in1len,
                       const unsigned char *in2, size_t in2len,
                       const unsigned char *nonce, size_t noncelen)
{
    RAND_DRBG_CTR *ctr = &drbg->ctr;

    /* ks is already setup for correct key */
    inc_128(ctr);
    AES_encrypt(ctr->V, ctr->K, &ctr->ks);

    /* If keylen longer than 128 bits need extra encrypt */
    if (ctr->keylen != 16) {
        inc_128(ctr);
        AES_encrypt(ctr->V, ctr->K + 16, &ctr->ks);
    }
    inc_128(ctr);
    AES_encrypt(ctr->V, ctr->V, &ctr->ks);

    /* If 192 bit key part of V is on end of K */
    if (ctr->keylen == 24) {
        memcpy(ctr->V + 8, ctr->V, 8);
        memcpy(ctr->V, ctr->K + 24, 8);
    }

    if (drbg->flags & RAND_DRBG_FLAG_CTR_USE_DF) {
        /* If no input reuse existing derived value */
        if (in1 != NULL || nonce != NULL || in2 != NULL)
            ctr_df(ctr, in1, in1len, nonce, noncelen, in2, in2len);
==>     /* If this a reuse input in1len != 0 */
==>     if (in1len)
==>         ctr_XOR(ctr, ctr->KX, drbg->seedlen);
    } else {
        ctr_XOR(ctr, in1, in1len);
        ctr_XOR(ctr, in2, in2len);
    }

    AES_set_encrypt_key(ctr->K, drbg->strength, &ctr->ks);
}
```
I don't understand the purpose of this and the cryptic comment `/* If this a reuse input in1len != 0 */` does not help much. That's why I chose the hack to be on the save side that reseeding works correctly.

The code is from the original FIPS object module, maybe @snhenson can shed some light on this.

Reply:
I'm sorry for giving you a hard time -- I meant that for the call to RAND_DRBG_reseed() on the line I was commenting on, adin will always be NULL and adinlen always 0.  The call to ctr_reseed() two lines previously looks fine to me.

At some point I thought I had convinced myself what the ```/* If this is a reuse input in1len != 0 */``` meant, namely that the combination of ```in1==NULL``` and ```in1len != 0``` was a specific sentinel that indicates this is an "reuse" case, but I don't remember exactly what was supposed to happen in the "reuse case", whether it was using a precached value from the structure or something else.

And, since it has not been said enough, thank you again, very much, for your patience and persistence working through this stuff -- I'm really glad that you are doing this work.

Reply:
> I meant that for the call to RAND_DRBG_reseed() on the line I was commenting on, adin will always be NULL and adinlen always 0. 

Yeah, I overlooked the obvious interpretation, because the `ctr_reseed()` problem distracted me. (I should also get used to looking at the bottom line of the source window attached to a comment ;-) ) But the good part about it is that it gave me the opportunity to write down my thoughts and concerns about it.

> At some point I thought I had convinced myself what the /* If this is a reuse input in1len != 0 */ meant, namely that the combination of in1==NULL and in1len != 0 was a specific sentinel that indicates this is an "reuse" case, but I don't remember exactly what was supposed to happen in the "reuse case", whether it was using a precached value from the structure or something else.

That's what I remember, too, although I was not able to find the location where this reuse occurs. (I didn't look very long.) In fact, I was worried more about the fact that the if-statement also applies to the non-reuse case and prevents me from passing `NULL` as `entropy` argument to `ctr_reseed()`. Now I am thinking about changing the condition to
```C
        if (in1len != 0 || in2len != 0)
            ctr_XOR(ctr, ctr->KX, drbg->seedlen);
```
such that I don't have to pass `adin` as fake `entropy` argument.

> And, since it has not been said enough, thank you again, very much, for your patience and persistence working through this stuff -- I'm really glad that you are doing this work.

And I am glad to have such careful reviewers. It gives me a better feeling about the PR. The random generator is a crucial part of the crypto library, and one has to be very, very careful not to mess anything up. After all, I don't want to be found guilty for introducing a RANDBLEED bug into OpenSSL in a few years.

Reply:
After having a final look at the `ctr_reseed()`, I decided not to change anything, because the call
```C
    ctr_reseed(drbg, adin, adinlen, NULL, 0);
```
is more or less equivalent to the call
```C
    ctr_update(drbg, adin, adinlen, NULL, 0, NULL, 0);
```
and the latter is used twice in `ctr_generate()`. So everything looks ok now to me, too.